src/include/main/client_context.h

```diff
@@ -138,26 +138,52 @@ class KUZU_API ClientContext {
 
     graph::GraphEntrySet& getGraphEntrySetUnsafe();
 
-    void cleanUP();
+    void cleanUp();
 
-    std::unique_ptr<QueryResult> queryInternal(std::string_view query,
+    // Query.
+    std::unique_ptr<PreparedStatement> prepare(std::string_view query);
+    std::unique_ptr<QueryResult> executeWithParams(PreparedStatement* preparedStatement,
+        std::unordered_map<std::string, std::unique_ptr<common::Value>> inputParams,
+        std::optional<uint64_t> queryID = std::nullopt);
+    std::unique_ptr<QueryResult> query(std::string_view queryStatement,
         std::optional<uint64_t> queryID = std::nullopt);
-
-    static std::string getEnvVariable(const std::string& name);
 
 private:
-    std::vector<std::shared_ptr<parser::Statement>> parseQuery(std::string_view query);
+    struct TransactionHelper {
+        enum class TransactionCommitAction : uint8_t {
+            COMMIT_IF_NEW,
+            COMMIT_IF_AUTO,
+            COMMIT_NEW_OR_AUTO,
+            NOT_COMMIT
+        };
+        static bool commitIfNew(TransactionCommitAction action) {
+            return action == TransactionCommitAction::COMMIT_IF_NEW ||
+                   action == TransactionCommitAction::COMMIT_NEW_OR_AUTO;
+        }
+        static bool commitIfAuto(TransactionCommitAction action) {
+            return action == TransactionCommitAction::COMMIT_IF_AUTO ||
+                   action == TransactionCommitAction::COMMIT_NEW_OR_AUTO;
+        }
+        static TransactionCommitAction getAction(bool commitIfNew, bool commitIfAuto);
+        static void runFuncInTransaction(transaction::TransactionContext& context,
+            const std::function<void()>& fun, bool readOnlyStatement, bool isTransactionStatement,
+            TransactionCommitAction action);
+    };
 
     static std::unique_ptr<QueryResult> queryResultWithError(std::string_view errMsg);
+    static std::unique_ptr<PreparedStatement> preparedStatementWithError(std::string_view errMsg);
+    static void bindParametersNoLock(const PreparedStatement* preparedStatement,
+        const std::unordered_map<std::string, std::unique_ptr<common::Value>>& inputParams);
+    void validateTransaction(const PreparedStatement& preparedStatement) const;
 
-    std::unique_ptr<PreparedStatement> preparedStatementWithError(std::string_view errMsg);
+    std::vector<std::shared_ptr<parser::Statement>> parseQuery(std::string_view query);
 
     // when we do prepare, we will start a transaction for the query
     // when we execute after prepare in a same context, we set requireNewTx to false and will not
     // commit the transaction in prepare when we only prepare a query statement, we set requireNewTx
     // to true and will commit the transaction in prepare
     std::unique_ptr<PreparedStatement> prepareNoLock(
-        std::shared_ptr<parser::Statement> parsedStatement, bool requireNewTx = true,
+        std::shared_ptr<parser::Statement> parsedStatement, bool shouldCommitNewTransaction = true,
```

Comment:
Let's not give default value to `shouldCommitNewTransaction` because it's not true to most function call will set it true. Also the comment is no longer accurate.